### PR TITLE
DebtCache refactoring

### PR DESCRIPTION
**Pull request by i-stam**
This PR follows the pattern introduced for splitting code in L1 and L2 counterparts by introducing a BaseDebtCache contract.
In this way: 
1.  Necessary functionality is explicitly stated instead of stubbing every function available.
2. It allows for further optimizations (code elimination and further code separation e.g. in Exchanger) 
3. It also allows for future decoupling/renaming. (there is no such thing as cache on L2)

### CODE REVIEW

File: contracts/BaseSynthetix.sol [hunk 1]
```diff
@@ -171,6 +171,22 @@ contract BaseSynthetix is IERC20, ExternStateToken, MixinResolver, ISynthetix {
         return exchanger().exchange(messageSender, sourceCurrencyKey, sourceAmount, destinationCurrencyKey, messageSender);
     }
 
+    function exchangeOnBehalf(
```
Comment (justinjmoses) [inline contracts/BaseSynthetix.sol:174]: is this change necessary?

Comment (justinjmoses) [reply contracts/BaseSynthetix.sol:174]: why move this function at all?

File: contracts/CollateralManager.sol [hunk 2]
```diff
@@ -15,7 +15,6 @@ import "./SafeDecimalMath.sol";
 import "./CollateralManagerState.sol";
 import "./interfaces/IIssuer.sol";
 import "./interfaces/IExchangeRates.sol";
-import "./interfaces/IDebtCache.sol";
```
Comment (justinjmoses) [inline contracts/CollateralManager.sol:18]: 💯

File: contracts/BaseDebtCache.sol [hunk 3]
```diff
@@ -0,0 +1,274 @@
+pragma solidity ^0.5.16;
+
+// Inheritance
+import "./Owned.sol";
+import "./MixinResolver.sol";
+import "./MixinSystemSettings.sol";
+import "./interfaces/IBaseDebtCache.sol";
```
Comment (justinjmoses) [inline contracts/BaseDebtCache.sol:7]: We can remove this interface

File: contracts/Exchanger.sol [hunk 4]
```diff
@@ -587,15 +586,17 @@ contract Exchanger is Owned, MixinSystemSettings, IExchanger {
...
-        appendExchange(
-            destinationAddress,
-            sourceCurrencyKey,
-            sourceAmountAfterSettlement,
-            destinationCurrencyKey,
-            amountReceived,
-            exchangeFeeRate
-        );
+        // persist the exchange information for the dest key iff the waiting period is gt 0
+        if (getWaitingPeriodSecs() > 0) {
```
Comment (justinjmoses) [inline contracts/Exchanger.sol:590]: this doesn't belong in this PR

File: publish/src/commands/deploy.js [hunk 5]
```diff
@@ -1846,7 +1846,7 @@ const deploy = async ({
 			read: 'waitingPeriodSecs',
 			expected: input => input !== '0',
 			write: 'setWaitingPeriodSecs',
-			writeArg: await getDeployParameter('WAITING_PERIOD_SECS'),
+			writeArg: useOvm ? '0' : await getDeployParameter('WAITING_PERIOD_SECS'),
```
Comment (justinjmoses) [inline publish/src/commands/deploy.js:1849]: as mentioned this shouldn't be here

File: contracts/Exchanger.sol (hunk 4 above)
Comment (i-stam) [reply contracts/Exchanger.sol:590]: yes, this happened because this branch was initially intended to serve as a PR for the one addressing the configurable waiting period issue.

File: publish/src/commands/deploy.js (hunk 5 above)
Comment (i-stam) [reply publish/src/commands/deploy.js:1849]: same problem as above, fixed!

File: contracts/BaseSynthetix.sol (hunk 1 above)
Comment (i-stam) [reply contracts/BaseSynthetix.sol:174]: This seemed to me as an extra functionality we can activate on L2, it doesn't have other dependencies like `exchangeWithVirtual()` for example.

File: contracts/BaseDebtCache.sol (hunk 3 above)
Comment (i-stam) [reply contracts/BaseDebtCache.sol:7]: Why should we remove it, all the basic debtCache functionality is described there. IDebtCache only declares the extra functionality needed i.e.`purgeCachedSynthDebt`, 'takeDebtSnapshot'.

Comment (justinjmoses) [reply contracts/BaseDebtCache.sol:7]: It isn't used anywhere? We only need `IDebtCache` in the system

File: contracts/Exchanger.sol [hunk 6]
```diff
@@ -587,7 +586,7 @@ contract Exchanger is Owned, MixinSystemSettings, IExchanger {
             destinationAddress
         );
 
-        // persist the exchange information for the dest key
+        // persist the exchange information for the dest key iff the waiting period is gt 0
```
Comment (justinjmoses) [inline contracts/Exchanger.sol:589]: [np] this comment is no longer accurate here

File: contracts/Issuer.sol [hunk 7]
```diff
@@ -23,7 +23,6 @@ import "./interfaces/IHasBalance.sol";
 import "./interfaces/IERC20.sol";
 import "./interfaces/ILiquidations.sol";
 import "./interfaces/ICollateralManager.sol";
-import "./interfaces/IDebtCache.sol";
```
Comment (justinjmoses) [inline contracts/Issuer.sol:26]: nice cleanup

**Review by justinjmoses — APPROVED** [review]
Nice one! Just a couple of minor points.
